### PR TITLE
rust/cargo: add support for a custom target directory

### PR DESCRIPTION
**Pull request by mathstuf**
This can avoid having to wait for ALE or ALE being blocked on other
cargo actions within the same crate.

---
> Where are the tests?

Where should I add them? I don't see any existing ones testing this kind of feature already.

Seriously, read `:help ale-dev` and write tests.

### CODE REVIEW

**Comment by mathstuf** [comment]
Ping @w0rp to un-stale this :) . It's not clear to me how to test this since it is controlling a `cargo` behavior.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -2,12 +2,12 @@
...
+* Read the documentation that comes with ALE with `:help ale-development`.
 
 Have fun!
 -->
 
 Where are the tests? Have you added tests? Have you updated the tests? Read the
 comment above and the documentation referenced in it first. Write tests!
 
-Seriously, read `:help ale-dev` and write tests.
+Seriously, read `:help ale-development` and write tests.
```
Comment (w0rp) [inline .github/PULL_REQUEST_TEMPLATE.md:13]: Undo these changes. The help tag is valid.

File: ale_linters/rust/cargo.vim [hunk 2]
```diff
@@ -82,6 +86,7 @@ function! ale_linters#rust#cargo#GetCommand(buffer, version) abort
     \   . (l:use_all_targets ? ' --all-targets' : '')
     \   . (l:use_examples ? ' --examples' : '')
     \   . (l:use_tests ? ' --tests' : '')
+    \   . (l:use_target_dir ? (' --target-dir ' . l:target_dir) : '')
```
Comment (w0rp) [inline ale_linters/rust/cargo.vim:89]: You need to quote paths with the `ale#Escape` function, which has cross-platform support.

```suggestion
    \   . (l:use_target_dir ? (' --target-dir ' . ale#Escape(l:target_dir)) : '')
```

Have a look at the `command_callback` tests documented in `:help ale-dev`. You can test the strings used for any command, and the versions we check. There will be number of tests for cargo already, so you can look at the current examples and add another test case.

File: ale_linters/rust/cargo.vim [hunk 3]
```diff
@@ -31,6 +32,9 @@ function! ale_linters#rust#cargo#GetCommand(buffer, version) abort
     \   && ale#semver#GTE(a:version, [0, 22, 0])
     let l:use_tests = ale#Var(a:buffer, 'rust_cargo_check_tests')
     \   && ale#semver#GTE(a:version, [0, 22, 0])
+    let l:use_target_dir = !empty(ale#Var(a:buffer, 'rust_cargo_target_dir'))
+    \   && ale#semver#GTE(a:version, [0, 17, 0])
+    let l:target_dir = ale#Var(a:buffer, 'rust_cargo_target_dir')
```
Comment (w0rp) [inline ale_linters/rust/cargo.vim:37]:
```suggestion
    let l:target_dir = ale#Var(a:buffer, 'rust_cargo_target_dir')
    let l:use_target_dir = !empty(l:target_dir)
    \   && ale#semver#GTE(a:version, [0, 17, 0])
```

You can-reuse the variable.

**Comment by mathstuf** [comment]
Thanks. Trying to run tests locally seems to hang with no CPU usage from the processes in the container...seeing what CI says.

**Comment by w0rp** [comment]
> Thanks. Trying to run tests locally seems to hang with no CPU usage from the processes in the container...seeing what CI says.

Yeah, I get that on my machine too. I haven't been able to figure out why the tests can do that, but if you just let them run for a while they will eventually exit and give you the results. At some point I'll try to figure out what is going on and fix that.

**Comment by w0rp** [comment]
The upshot of them taking a while with low CPU usage is that you can go off and do whatever else you want for a while, because your CPUs will be free to do other work.

**Comment by mathstuf** [comment]
Finally all green :) .

**Comment by w0rp** [comment]
Cheers! :beers:
